[WIP] Add documentation for Seccomp GA

hasheddan opened the pull request:
Signed-off-by: hasheddan <georgedanielmangum@gmail.com>

This is a WIP for adding documentation for the Seccomp GA KEP: https://github.com/kubernetes/enhancements/issues/135

Tasks still to be completed:
- [ ] Update `/content/en/docs/concepts/security/pod-security-standards.md`
- [ ] Update `/content/en/docs/tasks/configure-pod-container/security-context.md` with a short section on setting `seccompProfile` fields
- [ ] Update various examples through documentation that use seccomp annotations
- [ ] Add full guide in `/content/en/docs/tutorials/clusters/seccomp.md`

/cc @saschagrunert @pjbgf @evrardjp

sftim commented on content/en/docs/tutorials/clusters/seccomp.md:
```diff
@@ -0,0 +1,6 @@
+---
+reviewers:
+# - TODO 
+title: Seccomp
+content_template: templates/tutorial
+---
```
Early feedback: This would work better as a Task page, I think.

hasheddan replied:
Gotcha! Thanks :) I had just put it here because the `AppArmor` one was here as well, but I tend to agree with you

pjbgf commented on content/en/docs/concepts/policy/pod-security-policy.md:
```diff
@@ -599,28 +599,21 @@ documentation](/docs/tutorials/clusters/apparmor/#podsecuritypolicy-annotations)
+- `Localhost` - Specify a seccomp profile as a file on the node. Must set
+  `LocalhostProfile` if this option is provided.
+
+**LocalhostProfile** - Path of the seccomp profile in the format
   `<seccomp_root>/<path>`, where `<seccomp_root>` is defined via the
   `--seccomp-profile-root` flag on the Kubelet.
```
It may be worth mentioning that the flag `--seccomp-profile-root` is now deprecated and is due to be removed on v1.23.

pjbgf commented on content/en/docs/concepts/policy/pod-security-policy.md:
```diff
@@ -599,28 +599,21 @@ documentation](/docs/tutorials/clusters/apparmor/#podsecuritypolicy-annotations)
   `<seccomp_root>/<path>`, where `<seccomp_root>` is defined via the
   `--seccomp-profile-root` flag on the Kubelet.
 
-**seccomp.security.alpha.kubernetes.io/allowedProfileNames** - Annotation that
-specifies which values are allowed for the pod seccomp annotations. Specified as
-a comma-delimited list of allowed values. Possible values are those listed
-above, plus `*` to allow all profiles. Absence of this annotation means that the
-default cannot be changed.
-
```
This will need to be reverted.

saschagrunert commented on content/en/docs/concepts/policy/pod-security-policy.md:
```diff
@@ -599,28 +599,21 @@ documentation](/docs/tutorials/clusters/apparmor/#podsecuritypolicy-annotations)
 
-The use of seccomp profiles in pods can be controlled via annotations on the
-PodSecurityPolicy. Seccomp is an alpha feature in Kubernetes.
+`seccompProfile` contains two fields that specify the seccomp profile that will
+be used for the pod.
```
I suggest that we add a link to the new seccomp site here, too.

hasheddan replied:
@pjbgf  Ah great point! I misread what was being described here :/ I'll revert changes and remove the alpha specification :+1:

hasheddan replied:
@saschagrunert by site do you mean the full guide that will be in `/tasks`?

saschagrunert replied:
> @saschagrunert by site do you mean the full guide that will be in `/tasks`?

Yes :+1: :)

savitharaghunathan commented:
Hi @hasheddan, Thank you for the PR. If this is a part of release 1.19, which I assume it is, can you open this PR against `dev-1.19` branch, please? Thanks!

hasheddan commented:
@savitharaghunathan sure thing! Thanks for letting me know!

sftim commented:
/milestone 1.19
AIUI

sftim commented:
/sig node

sftim commented on content/en/docs/concepts/policy/pod-security-policy.md:
```diff
@@ -602,7 +602,7 @@ documentation](/docs/tutorials/clusters/apparmor/#podsecuritypolicy-annotations)
 The use of seccomp profiles in pods can be controlled via annotations on the
```
```suggestion
You can use annotations on Pods to control use of seccomp profiles, and you can use
PodSecurityPolicies to enforce how these annotations are applied.
```

savitharaghunathan commented:
/unassign @rajakavitha1
/assign

hasheddan commented:
@sftim I believe that `--seccomp-profile-root` will continue to be honored until it is removed in v1.23. @pjbgf @saschagrunert  could you verify this?

saschagrunert commented on content/en/docs/concepts/services-networking/network-policies.md:
```diff
@@ -205,9 +205,9 @@ This ensures that even pods that aren't selected by any other NetworkPolicy will
 ## SCTP support
 
-{{< feature-state for_k8s_version="v1.12" state="alpha" >}}
+{{< feature-state for_k8s_version="v1.19" state="beta" >}}
```
I'm wondering why this is part of this PR. What was the reason for it?

hasheddan replied:
@saschagrunert oh weird, looks like I messed up the rebase. None of these SCTP changes are mine

hasheddan replied:
Should be fixed now :+1:

tengqm commented:
still a WIP?
/lgtm

pjbgf commented on content/en/docs/concepts/policy/pod-security-policy.md:
```diff
@@ -599,23 +599,26 @@ documentation](/docs/tutorials/clusters/apparmor/#podsecuritypolicy-annotations)
 
-The use of seccomp profiles in pods can be controlled via annotations on the
-PodSecurityPolicy. Seccomp is an alpha feature in Kubernetes.
+You can use annotations on Pods to control use of seccomp profiles, and you can
+use PodSecurityPolicies to enforce how these annotations are applied.
```
May be worth focusing on secomp GA behaviour:

```suggestion
You can use the seccompProfile field on Pods or containers to control use of seccomp profiles, 
and you can use PodSecurityPolicies to enforce how these fields are applied.
```

pjbgf commented on content/en/examples/policy/privileged-psp.yaml:
```diff
@@ -3,7 +3,7 @@ kind: PodSecurityPolicy
   name: privileged
   annotations:
-    seccomp.security.alpha.kubernetes.io/allowedProfileNames: '*'
+    seccomp.security.kubernetes.io/allowedProfileNames: '*'
```
@hasheddan @saschagrunert The KEP is not explicit on this point, but I thought we were keeping the PSP seccomp feature in alpha as we were [keeping PSP functionality unchanged](https://github.com/kubernetes/enhancements/pull/1747#issuecomment-631105776). I am unsure whether it is worth getting users to "migrate" from the existing annotation to one without `alpha`.

hasheddan replied:
Yeah I think you are right. It does feel a little strange that this is `alpha` while seccomp itself is GA, but this can be changed in a future release if needed :+1:

pjbgf replied:
Adding @tallclair @liggitt who were involved in the discussions around the subject.

hasheddan commented:
@saschagrunert @pjbgf I have added the tutorial section and addressed comments :) WIP designation has been removed and this PR is ready for final review :heavy_check_mark: 

/cc @savitharaghunathan

pjbgf commented on content/en/docs/concepts/security/pod-security-standards.md:
```diff
@@ -249,13 +249,14 @@ well as lower-trust users.The following listed controls should be enforced/disal
 			<td>Seccomp</td>
 			<td>
-				The 'runtime/default' seccomp profile must be required, or allow specific additional profiles.<br>
+				The RuntimeDefault seccomp profile must be required, or allow specific additional profiles..<br>
```
```suggestion
				The RuntimeDefault seccomp profile must be required, or allow specific additional profiles.<br>
```

pjbgf commented on content/en/docs/concepts/security/pod-security-standards.md:
```diff
@@ -249,13 +249,14 @@ well as lower-trust users.The following listed controls should be enforced/disal
+				spec.initContainers[*].securityContext.seccompProfile.type<br>
 				<br><b>Allowed Values:</b><br>
 				'runtime/default'<br>
-				undefined (container annotation)<br>
+				undefined / nil<br>
```
Considering `.type`, would the allowed values not be: `RuntimeDefault`, `Unconfined` and `Localhost`? If we refer to a level up (a.k.a `.seccompProfile`) then we can probably just leave `undefined/nil`.

hasheddan replied:
Hmm good point. Yeah I think what we are trying to say here (though I feel this document is a bit confusing) is that you can skip adding a `securityContext.seccomp` section to your containers. To communicate this do you just think we should remove `.type` off of the `initContainers` and `containers` lines?

pjbgf replied:
yep, I think if we just remove `.type` it becomes clear enough. :-1:

saschagrunert commented on content/en/docs/concepts/policy/pod-security-policy.md:
```diff
@@ -599,8 +599,10 @@ documentation](/docs/tutorials/clusters/apparmor/#podsecuritypolicy-annotations)
+You can use the `seccompProfile` field in the `securityContext` of Pods or
+containers to [control use of seccomp
+profiles](../../tutorials/clusters/seccomp.md), and you can use
+PodSecurityPolicies to enforce how these fields are applied.
```
Let's mention that the PSP enforces both: the annotation and the new fields like described here: https://github.com/kubernetes/enhancements/blob/master/keps/sig-node/20190717-seccomp-ga.md#podsecuritypolicy-enforcement

hasheddan replied:
Great point, thanks @saschagrunert!

hasheddan replied:
Updated :+1:

pjbgf commented:
/lgtm

hasheddan commented:
@savitharaghunathan @sftim are additional steps needed here? I tried to check how this renders locally, but am having issues serving in container:

```
ERROR 2020/07/09 17:49:39 render of "page" failed: execute of template failed: template: case-studies/single.html:4:5: executing "case-studies/single.html" at <partial "head.html" .>: error calling partial: "/src/layouts/partials/head.html:35:3": execute of template failed: template: partials/head.html:35:3: executing "partials/head.html" at <partialCached "head-css.html" . "asdf">: error calling partialCached: partial "head-css.html" not found
ERROR 2020/07/09 17:49:39 render of "page" failed: execute of template failed: template: case-studies/single.html:4:5: executing "case-studies/single.html" at <partial "head.html" .>: error calling partial: "/src/layouts/partials/head.html:35:3": execute of template failed: template: partials/head.html:35:3: executing "partials/head.html" at <partialCached "head-css.html" . "asdf">: error calling partialCached: partial "head-css.html" not found
ERROR 2020/07/09 17:49:39 render of "page" failed: execute of template failed: template: case-studies/single.html:4:5: executing "case-studies/single.html" at <partial "head.html" .>: error calling partial: "/src/layouts/partials/head.html:35:3": execute of template failed: template: partials/head.html:35:3: executing "partials/head.html" at <partialCached "head-css.html" . "asdf">: error calling partialCached: partial "head-css.html" not found
ERROR 2020/07/09 17:49:39 render of "page" failed: execute of template failed: template: case-studies/single.html:4:5: executing "case-studies/single.html" at <partial "head.html" .>: error calling partial: "/src/layouts/partials/head.html:35:3": execute of template failed: template: partials/head.html:35:3: executing "partials/head.html" at <partialCached "head-css.html" . "asdf">: error calling partialCached: partial "head-css.html" not found
Built in 1662 ms
Error: Error building site: failed to render pages: render of "page" failed: execute of template failed: template: case-studies/single.html:4:5: executing "case-studies/single.html" at <partial "head.html" .>: error calling partial: "/src/layouts/partials/head.html:35:3": execute of template failed: template: partials/head.html:35:3: executing "partials/head.html" at <partialCached "head-css.html" . "asdf">: error calling partialCached: partial "head-css.html" not found
make: *** [Makefile:70: container-serve] Error 255
```

sftim commented:
@hasheddan 

1. Try rebasing against _dev-1.19_ and then try again.
2. Local previewing has changed. This should work (on POSIX-like systems):
   ```bash
   cd "$( git rev-parse --show-toplevel )" && \
   git submodule update --init --recursive --depth 1 && make container-serve
   ```

hasheddan commented:
@sftim Thanks! That worked great. I have corrected a few links that were broken. @saschagrunert @pjbgf can yall circle back around here for `lgtm` ?

pjbgf commented:
/lgtm

annajung commented:
Hi @hasheddan, docs shadow here 👋. Just a friendly reminder that this PR needs tech and doc review by July 16th. 

Looks like the technical review is done and just waiting for review from docs. 
@sftim, Can I ask you to re-review? pinging you since you're active on this PR :)

kbhawkey commented:
@tallclair , Would you have time to review these changes? Thanks!

celestehorgan commented:
/assign

sftim commented on content/en/docs/tutorials/clusters/seccomp.md:
```diff
@@ -0,0 +1,346 @@
+- hasheddan
+- pjbgf
+- saschagrunert
+title: Seccomp
```
```suggestion
title: Isolate a workload using Linux Seccomp
```
?

hasheddan replied:
This was to match the existing `AppArmor` title.

sftim commented on content/en/docs/tutorials/clusters/seccomp.md:
```diff
@@ -0,0 +1,346 @@
+
+In order to complete all steps in this tutorial, you must install
+[kind](https://kind.sigs.k8s.io/docs/user/quick-start/) and
+[kubectl](../../tasks/tools/install-kubectl). We will show examples with both
```
```suggestion
[kubectl](/doc/tasks/tools/install-kubectl/). This tutorial both
```
([avoid using “we”](https://kubernetes.io/docs/contribute/style/style-guide/#avoid-using-we))

sftim commented on content/en/docs/tutorials/clusters/seccomp.md:
```diff
@@ -0,0 +1,346 @@
+In order to complete all steps in this tutorial, you must install
+[kind](https://kind.sigs.k8s.io/docs/user/quick-start/) and
+[kubectl](../../tasks/tools/install-kubectl). We will show examples with both
+`alpha` (pre-v1.19) and GA seccomp functionality, so make sure that your cluster
```
alpha (pre-v1.19) and generally available seccomp functionality, so make sure that your cluster

kbhawkey replied:
@hasheddan , Why is it important to compare the `alpha` seccomp with `GA` seccomp?
It makes the pathway through the tutorial more complicated.

hasheddan replied:
@kbhawkey If someone is using a Kubernetes version pre-v1.19 the GA behavior will not be available. Because it takes time for people to update to newer versions of Kubernetes I feel like it makes sense to demonstrate how seccomp can be used in both settings.

sftim replied:
@hasheddan you might not know that Kubernetes documentation is also available for previous versions.

In the documentation for v1.19 and later you can tell readers to look back at the documentation for previous releases - eg https://v1-18.docs.kubernetes.io/ (and this can be different text).
That's how I'd do it.

If you want to achieve this we can explain how.

hasheddan replied:
@sftim I have implemented the tabs approach as suggested in https://github.com/kubernetes/website/pull/21278#discussion_r454694898. This feels like a clean experience to me, but I am happy to remove the alpha sections and make a separate revision of this tutorial for 1.18 if needed. Let me know what you think

sftim commented on content/en/docs/tutorials/clusters/seccomp.md:
```diff
@@ -0,0 +1,346 @@
+
+## Create Seccomp Profiles
+
+We will discuss the contents of these profiles later on, but for now go ahead
+and download them into a directory named `profiles/` so that we can load them
+into our cluster.
```
nit: reword to avoid using “we”

kbhawkey replied:
Note: The json code blocks are rendering side-by-side depending on horizontal size of page.
Need to log issue.

sftim commented on content/en/docs/tutorials/clusters/seccomp.md:
````diff
@@ -0,0 +1,346 @@
+audit.json  fine-grained.json  violation.json
+```
+
+## Create Pod with Seccomp Profile for Syscall Auditing
````
nit:
```suggestion
## Create a Pod with a Seccomp profile for syscall auditing
```

sftim commented on content/en/docs/tutorials/clusters/seccomp.md:
```diff
@@ -0,0 +1,346 @@
+
+{{< codenew file="pods/security/seccomp/ga/audit-pod.yaml" >}}
+
+Create the Pod in your cluster.
```
nit:
```suggestion
Create the Pod in your cluster:
```

sftim commented on content/en/docs/concepts/policy/pod-security-policy.md:
```diff
@@ -609,11 +612,14 @@ are:
 - `localhost/<path>` - Specify a profile as a file on the node located at
   `<seccomp_root>/<path>`, where `<seccomp_root>` is defined via the
-  `--seccomp-profile-root` flag on the Kubelet.
+  `--seccomp-profile-root` flag on the Kubelet. (Note: this flag has been
+  deprecated in v1.19 and will be removed in v1.23. The seccomp root path will
+  then be derived from the kubelet root path, which is defined by `--root-dir`.
+  The current default value is `<root-dir>/seccomp`.)
```
```suggestion
  `--seccomp-profile-root` flag on the kubelet.

   {{< note >}}
   Kubernetes v1.19 deprecated the `--seccomp-profile-root` flag, which will be removed in
   Kubernetes v1.23. The seccomp root path will   then be derived from the kubelet root path,
   which is defined by `--root-dir`. The current default value is `<root-dir>/seccomp`.
   {{< /note >}}
```
_or_, with [avoid statements about the future](https://kubernetes.io/docs/contribute/style/style-guide/#avoid-statements-about-the-future) in mind:
```suggestion
   the seccomp root path for that node. The default seccomp root path is a directory
   named `seccomp` directly beneath the kubelet's root directory (specified via the
   `--root-dir` command line flag to the kubelet).

   {{< note >}}
   The `--seccomp-profile-root` flag (deprecated since Kubernetes v1.19) allowed you to configure 
   the seccomp root path for a node. To use the defined behavior, leave that flag unset so that the
   kubelet uses its default.   
   {{< /note >}}
```

kbhawkey replied:
@hasheddan , avoid statements/promises/predictions about the future.

sftim commented on content/en/docs/tutorials/clusters/apparmor.md:
```diff
@@ -3,6 +3,7 @@ reviewers:
 - stclair
 title: AppArmor
```
Consider revising this title in the PR that introduces a Seccomp tutorial.

hasheddan commented:
@sftim how would opening a separate PR work with docs freeze? This PR was marked ready for review before the deadline but a new one would not have been.

kbhawkey commented:
Page previews:
https://deploy-preview-21278--kubernetes-io-vnext-staging.netlify.app/docs/tutorials/clusters/seccomp/
https://deploy-preview-21278--kubernetes-io-vnext-staging.netlify.app/docs/tutorials/clusters/apparmor/
https://deploy-preview-21278--kubernetes-io-vnext-staging.netlify.app/docs/concepts/security/pod-security-standards/
https://deploy-preview-21278--kubernetes-io-vnext-staging.netlify.app/docs/concepts/policy/pod-security-policy/#seccomp

sftim commented:
> @sftim how would opening a separate PR work with docs freeze?

@hasheddan the website itself uses a continuous deployment pipeline. To get your change into a release (v1.19) you must document it, and this documentation has an impending deadline for when your feature documentation must be ready. But, you can split this PR into a minimum-viable amount of documentation and a separate PR with the tutorial in it, both targeting _dev-1.19_.

Once #21278 (this PR) merges, and a putative separate PR for adding the seccomp tutorial has */lgtm*, then that separate tutorial PR is good to merge into _dev-1.19_, ahead of the release. If the separate tutorial misses the opportunity to merge into the dev-1.19 branch you can change its merge target to _master_, after v1.19 is released, and it'll still be ready for approval, automated merge, and publishing.

Essentially: whilst it's nice to add that tutorial for using Seccomp, not adding that tutorial doesn't block or delay the v1.19 release.

@savitharaghunathan & @kubernetes/sig-docs-leads does that sound right? If I've got a detail wrong there please jump in.

kbhawkey commented on content/en/docs/tutorials/clusters/seccomp.md:
```diff
@@ -0,0 +1,346 @@
+## {{% heading "prerequisites" %}}
+
+In order to complete all steps in this tutorial, you must install
+[kind](https://kind.sigs.k8s.io/docs/user/quick-start/) and
```
Is it straightforward to port the tutorial to use `minikube`?

kbhawkey commented on content/en/docs/tutorials/clusters/seccomp.md:
```diff
@@ -0,0 +1,346 @@
+
+## Create Pod with Seccomp Profile for Syscall Auditing
+
+Next we want to apply our `audit.json` profile, which will log all syscalls of
```
Next , apply the `audit.json` profile. The `audit.json` profile logs all syscalls of the process to a new Pod.

sftim commented on content/en/docs/tutorials/clusters/seccomp.md:
```diff
@@ -0,0 +1,346 @@
+Next we want to apply our `audit.json` profile, which will log all syscalls of
+the process, to a new Pod. 
+
+If you are using a pre-v1.19 Kubernetes version, download this manifest as
+`audit-pod.yaml`. 
+
+{{< codenew file="pods/security/seccomp/alpha/audit-pod.yaml" >}}
+
+If using Kubernetes v1.19 or later, download this manifest as `audit-pod.yaml`.
+
+{{< codenew file="pods/security/seccomp/ga/audit-pod.yaml" >}}
```
Consider using [tabs](https://kubernetes.io/docs/contribute/style/hugo-shortcodes/#tabs) here.

sftim replied:
A viable alternative: get a version of this tutorial for v1.18 merged into _master_, and then revise it for v1.19

In the v1.19 and later versions, tell the reader to look at the v1.18 version if they're using an older cluster.

kbhawkey commented on content/en/docs/tutorials/clusters/seccomp.md:
````diff
@@ -0,0 +1,346 @@
+audit-pod   1/1     Running   0          30s
+```
+
+We want to be able to interact with this endpoint exposed by this container, so
````
nit, clean up `we` and `let's`:
To interact with this endpoint exposed by this container, create a NodePort service that allows access
to the endpoint from inside your `kind` control plane container.

kbhawkey commented on content/en/docs/tutorials/clusters/seccomp.md:
````diff
@@ -0,0 +1,346 @@
+```
+
+Now we will be able to `curl` the endpoint from inside the kind control plane
+container at the port exposed by this Service.
````
Clean up: `Now we will be able` ...

kbhawkey commented on content/en/docs/tutorials/clusters/seccomp.md:
````diff
@@ -0,0 +1,346 @@
+```
+docker exec -it 6a96207fed4b curl localhost:32373
+```
+
````
Is this the output?
If so, could add text before the code block to explain.

hasheddan replied:
This is a command, added some text to indicate so :+1:

kbhawkey commented on content/en/docs/tutorials/clusters/seccomp.md:
```diff
@@ -0,0 +1,346 @@
+
+{{< codenew file="pods/security/seccomp/alpha/default-pod.yaml" >}}
+
+If using Kubernetes v1.19 or later, set the field.
```
nit: Set the field?

kbhawkey commented:
Looking at this again.

kbhawkey commented:
Hello @hasheddan . Thanks for this PR.
I looked through the changes again. If you have time, you could fix the minor style issues.
I don't have a strong opinion about which cluster setup tool to feature with the examples.
I'd be in favor of showing both `kind` and `minikube` paths. I am okay with showing `kind` now and updating later.
For me, the biggest challenge with this tutorial is displaying the `alpha` and `GA` manifests together.
How difficult would it be to split the tutorial, creating two versions (alpha, GA) of the tutorial?

kbhawkey commented:
@savitharaghunathan , What is the absolute last day/time for accepting 1.19 docs changes?

savitharaghunathan commented:
> @savitharaghunathan , What is the absolute last day/time for accepting 1.19 docs changes?

@kbhawkey It is EOD PST time today :) However, this particular PR has obtained approvals to be merged after the deadline.

hasheddan commented:
@sftim @kbhawkey thank you for your thoughtful reviews and your patience! I believe I have now addressed all comments, and have made the explanation of using both GA and alpha seccomp cleaner by using tabs. I have not yet moved the tutorial from kind to minikube, and would advocate for keeping kind and adding minikube because I think the kind setup is useful information. Seccomp should work on any Kubernetes cluster where it is enabled, so it does not make a huge difference what tool or service is used. I am starting work on adding minikube support here, but please let me know if you think it is not necessary. Thanks again for your work!

hasheddan commented:
@zacharysarah commits squashed :+1:

zacharysarah commented:
/lgtm
/approve
